File: sdk/typescript/src/profitability.ts
/**
 * VAEA Flash — Profitability Calculator
 *
 * Calculates if a flash loan strategy is profitable after all fees.
 * Uses data from our own getQuote() API — zero external dependencies.
 */

import { QuoteResponse } from './types';

export interface ProfitabilityParams {
  /** Token symbol or mint */
  token: string;
  /** Borrow amount in human units */
  amount: number;
  /** Expected gross profit from strategy in SOL (e.g. arb profit after repay) */
  expectedRevenue: number;
  /** Optional Jito tip in SOL */
  jitoTip?: number;
  /** Optional priority fee in SOL */
  priorityFee?: number;
}

export interface ProfitabilityResult {
  /** Whether the strategy is net-profitable */
  profitable: boolean;
  /** Net profit in token units (negative = loss) */
  netProfit: number;
  /** Detailed cost breakdown */
  breakdown: {
    revenue: number;
    vaeFee: number;
    swapFees: number;
    networkFee: number;
    priorityFee: number;
    jitoTip: number;
    totalCost: number;
  };
  /** 'send' if profit > 2x costs, 'wait' if marginal, 'abort' if unprofitable */
  recommendation: 'send' | 'wait' | 'abort';
}

/**
 * Calculate profitability of a flash loan strategy.
 *
 * @param getQuote - Function to fetch a quote (bound from VaeaFlash instance)
 * @param params - Strategy parameters
 * @returns Detailed profitability analysis
 */
export async function calculateProfitability(
  getQuote: (token: string, amount: number) => Promise<QuoteResponse>,
  params: ProfitabilityParams,
): Promise<ProfitabilityResult> {
  const quote = await getQuote(params.token, params.amount);

  const vaeFee = quote.fee_breakdown.vaea_fee;
  const swapFees = quote.fee_breakdown.swap_in_fee + quote.fee_breakdown.swap_out_fee;
  const networkFee = 0.000005; // ~5000 lamports base fee (fixed on Solana)
  const priorityFee = params.priorityFee ?? 0;
  const jitoTip = params.jitoTip ?? 0;

  const totalCost = vaeFee + swapFees + networkFee + priorityFee + jitoTip;
  const netProfit = params.expectedRevenue - totalCost;

  let recommendation: 'send' | 'wait' | 'abort';
  if (netProfit > totalCost * 2) {
    recommendation = 'send';    // Clear profit: > 2x costs
  } else if (netProfit > 0) {
    recommendation = 'wait';    // Marginal profit
  } else {
    recommendation = 'abort';   // Unprofitable
  }

  return {
    profitable: netProfit > 0,
    netProfit,
    breakdown: {
      revenue: params.expectedRevenue,
      vaeFee,
      swapFees,
      networkFee,
      priorityFee,
      jitoTip,
      totalCost,
    },
    recommendation,
  };
}

File: sdk/rust/src/profitability.rs
/// VAEA Flash — Profitability Calculator (Rust)
///
/// Calculates if a flash loan strategy is profitable after all fees.
/// Uses data from our own QuoteResponse — zero external dependencies.

use crate::types::QuoteResponse;

/// Parameters for profitability check.
pub struct ProfitabilityParams {
    /// Token symbol or mint
    pub token: String,
    /// Borrow amount in human units
    pub amount: f64,
    /// Expected gross profit from strategy in SOL (e.g. arb profit after repay)
    pub expected_revenue: f64,
    /// Optional Jito tip in SOL
    pub jito_tip: f64,
    /// Optional priority fee in SOL
    pub priority_fee: f64,
}

/// Profitability analysis result.
pub struct ProfitabilityResult {
    /// Whether the strategy is net-profitable
    pub profitable: bool,
    /// Net profit in SOL (negative = loss)
    pub net_profit: f64,
    /// Detailed cost breakdown
    pub breakdown: ProfitabilityBreakdown,
    /// Recommendation: "send" if profit > 2x costs, "wait" if marginal, "abort" if unprofitable
    pub recommendation: Recommendation,
}

pub struct ProfitabilityBreakdown {
    pub revenue: f64,
    pub vaea_fee: f64,
    pub swap_fees: f64,
    pub network_fee: f64,
    pub priority_fee: f64,
    pub jito_tip: f64,
    pub total_cost: f64,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Recommendation {
    /// Clear profit (> 2x costs)
    Send,
    /// Marginal profit
    Wait,
    /// Unprofitable
    Abort,
}

/// Calculate profitability from a QuoteResponse.
pub fn calculate_profitability(
    quote: &QuoteResponse,
    params: &ProfitabilityParams,
) -> ProfitabilityResult {
    let vaea_fee = quote.fee_breakdown.vaea_fee;
    let swap_fees = quote.fee_breakdown.swap_in_fee + quote.fee_breakdown.swap_out_fee;
    let network_fee = 0.000005; // ~5000 lamports base fee
    let priority_fee = params.priority_fee;
    let jito_tip = params.jito_tip;

    let total_cost = vaea_fee + swap_fees + network_fee + priority_fee + jito_tip;
    let net_profit = params.expected_revenue - total_cost;

    let recommendation = if net_profit > total_cost * 2.0 {
        Recommendation::Send
    } else if net_profit > 0.0 {
        Recommendation::Wait
    } else {
        Recommendation::Abort
    };

    ProfitabilityResult {
        profitable: net_profit > 0.0,
        net_profit,
        breakdown: ProfitabilityBreakdown {
            revenue: params.expected_revenue,
            vaea_fee,
            swap_fees,
            network_fee,
            priority_fee,
            jito_tip,
            total_cost,
        },
        recommendation,
    }
}
